source/HardwareSerial.cpp

```diff
@@ -31,276 +63,559 @@ HardwareSerial::~HardwareSerial()
-    { sizeof(DCB), 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, ODDPARITY, ONESTOPBIT, 17, 19, 0, 0, 0, 0 },
-    // SERIAL_8O1
-    { sizeof(DCB), 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 8, ODDPARITY, ONESTOPBIT, 17, 19, 0, 0, 0, 0 },
-    // SERIAL_5O2
-    { sizeof(DCB), 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, ODDPARITY, TWOSTOPBITS, 17, 19, 0, 0, 0, 0 },
-    // SERIAL_6O2
-    { sizeof(DCB), 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, ODDPARITY, TWOSTOPBITS, 17, 19, 0, 0, 0, 0 },
-    // SERIAL_7O2
-    { sizeof(DCB), 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 7, ODDPARITY, TWOSTOPBITS, 17, 19, 0, 0, 0, 0 },
-    // SERIAL_8O2
-    { sizeof(DCB), 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 8, ODDPARITY, TWOSTOPBITS, 17, 19, 0, 0, 0, 0 }
-};
-
 HardwareSerial::operator bool(void)
 {
-    if (_comHandle != NULL && _comHandle != INVALID_HANDLE_VALUE)
-    {
-        return true;
-    }
-    return false;
+    return true;
 }
 
 int HardwareSerial::available(void)
 {
-    if ((_storageCount - _storageIndex) <= 0)
+    std::deque<IBuffer^>::const_iterator iter;
+    int bytesAvailable = 0;
+
+    EnterCriticalSection(&m_readBufferListLock);
+    iter = m_readBufferList.begin();
+    while (iter != m_readBufferList.end())
     {
-        peek();
+        bytesAvailable += (*iter)->Length;
+        ++iter;
     }
-    return _storageCount - _storageIndex;
+    LeaveCriticalSection(&m_readBufferListLock);
+
+    return bytesAvailable;
 }
 
 void HardwareSerial::setTimeout(unsigned long timeout)
 {
-    _timeout = timeout;
-
-    // Sets WIN32 READ/WRITE timeouts
-    COMMTIMEOUTS CommTimeouts;
-    GetCommTimeouts(_comHandle, &CommTimeouts);
+    m_timeout = timeout;
+}
 
-    CommTimeouts.ReadIntervalTimeout = 0;
-    CommTimeouts.ReadTotalTimeoutConstant = _timeout;
-    CommTimeouts.ReadTotalTimeoutMultiplier = 0;
+void HardwareSerial::begin(unsigned long baud, uint8_t config)
+{
+    HRESULT hr = S_OK;
 
-    CommTimeouts.WriteTotalTimeoutConstant = _timeout;
-    CommTimeouts.WriteTotalTimeoutMultiplier = 0;
+    hr = g_pins.verifyPinFunction(m_txPinNo, FUNC_SER, BoardPinsClass::LOCK_FUNCTION);
+    if (FAILED(hr))
+    {
+        ThrowError(hr, "An error occurred configuring pin %d as for Serial TX use.", m_txPinNo);
+    }
 
-    if (!SetCommTimeouts(_comHandle, &CommTimeouts))
+    hr = g_pins.verifyPinFunction(m_rxPinNo, FUNC_SER, BoardPinsClass::LOCK_FUNCTION);
+    if (FAILED(hr))
     {
-#ifdef _DEBUG
-        Log("Error %d when setting Com timeouts: ", GetLastError());
-        LogLastError();
-#endif
+        ThrowError(hr, "An error occurred configuring pin %d for Serial RX use.", m_rxPinNo);
     }
-}
 
-void HardwareSerial::begin(unsigned long baud, uint8_t config)
-{
-    DCB dcb = dcbArray[config];
+    // Set up objects needed to cancel read threads when .end() is called.
+    m_cancellationTokenSource = new Concurrency::cancellation_token_source();
```

Comment:
I figured we would allow the throw.  If we don't have the lock, we would end up failing and throwing an exception later anyway.